### DISCUSSION TITLE
Brier Skill Score Constant Reference Climatology

### DISCUSSION THREAD

**Discussion by jhicks14** [dtcenter/METplus#1635] [Configuration] [answered]
I'm having issues using METplus (v4.0) to define a constant reference climatology for use in the BSS calculation; e.g., 0.5 for a 2-category system. My PSTD output file shows up with "NA" for BRIERCL and BSS, but the BSS_SMPL is present. 

I've tried reading in a mean climatology file with constant grids set to 0.5, but that did not seem to work. I've also tried setting the "GRID_STAT_CLIMO_CDF_DIRECT_PROB" to "false" and setting the "GRID_STAT_CLIMO_CDF_BINS" to 2 (for a 2-cat system), but I still get "NA" in the PSTD output.

I'm attaching my configuration file for reference. It contains quite a few different lines that I may or may not need, but I have left them in to keep track of my attempts to get this to work.  

[Grid_SeriesAnl_Global_SubX_MP4_GEFSv12_gh500_BSS.txt](https://github.com/dtcenter/METplus/files/8788425/Grid_SeriesAnl_Global_SubX_MP4_GEFSv12_gh500_BSS.txt)

**Comment by CPKalb** [comment]
Hi:

Would you be able to send us a model and observation file, so we can take a closer look?  Instructions for how to send data can be found [here](https://github.com/dtcenter/METplus/discussions/954).

Thanks,
Christina

**Reply by jhicks14** [reply]
@CPKalb Just sent the data files! Should be under "Hicks_data"

**Reply by CPKalb** [reply]
Thanks.  Is your constant climatology given as a probability?

Christina

**Reply by jhicks14** [reply]
The constant climo field is set to 0.5 to represent a 50% probability

**Comment by CPKalb** [comment]
Hi:

If I set the log verbosity up on a METplus run, it looks like it's not reading the climatology data.  However, I'm not sure yet why it's not reading the data.  I'll have to dig a bit more.

warnings from MET:
WARNING:
WARNING: Dictionary::lookup_array() -> array lookup failed for name "climo_mean.field"
WARNING:
WARNING:
WARNING: Dictionary::lookup_array() -> array lookup failed for name "climo_stdev.field"
WARNING:

Christina

**Reply by jhicks14** [reply]
@CPKalb Thanks for looking into this!

**Comment by CPKalb** [comment]
Hi:

After fixing the data read, METplus is still not using the climatology field.  I get the following error:
 Skipping the 20140101_000000 "tmean(0,*,*)" climatology field since the time offset (9590400 seconds) >= the "day_interval" entry (2678400 seconds) from file: climo_BSS_fixed.nc

The time offset when comparing the date on the climatology data is outside the default time range so it is not used.  This time check can be turned off by setting day_interval = NA in the climatology mean dictionary in a MET configuration file.  In METplus, this should be controlled by GRID_STAT_CLIMO_MEAN_DAY_INTERVAL.  However, setting this variable to NA results in an error, as it seems METplus only accepts numbers.   This is something we need to fix. 

I was, however, able to get BSS and BRIERCL by using the config overrides.  To do this, change the following line in your config file:
GRID_STAT_MET_CONFIG_OVERRIDES =  climo_mean = {field = [{name="tmean"; level="(0,*,*)";}]; day_interval = NA;};

With this change, I get BSS and BRIERCL output in the pstd file.  In this case, you shouldn't need the climo_stdev field since you climatology data is already given as a probability.   Does this work on your end?

Thanks,
Christina

**Reply by jhicks14** [reply]
@CPKalb Thanks! That override did help me get the BSS and BRIERCL to output in the pstd file. However, the BRIERCL value is not the constant "0.5" value that I expected it to be based on the climo_BSS_fixed.nc file (which just contains a constant field of 0.5 for every grid point).

**Reply by CPKalb** [reply]
The BRIERCL is the Brier Score computed on the climatology.  Observations are used in this calculation, so it likely won't be equal to the constant value.

**Reply by jhicks14** [reply]
Perhaps this is just a different calculation of Brier Score than I am used to? 

My thought was that the brier score would be calculated as (1-0.5)^2, and leave me a BRIERCL score of 0.25. 

Rather, if my constant climo field is set to 0.5, I get a BRIERCL score of 0.26814 for the CONUS mask. A constant climo field set to 0.1 yields a BRIERCL score of 0.26303 for the CONUS mask. These values change depending on the region of interest.

**Reply by CPKalb** [reply]
MET's Brier score calculation can be found [here](https://met.readthedocs.io/en/latest/Users_Guide/appendixC.html#met-verification-measures-for-probabilistic-forecasts) in section 32.4.4.  It's a little more complex than the simple version that many have seen and computes Brier using a contingency table as opposed to using individual probability forecasts.

**Reply by jhicks14** [reply]
@CPKalb I see! Is there any way to ensure that the midpoints of the binned probability values are being used?

**Comment by j-opatz** [comment, marked as answer]
Hi @jhicks14,

I know we discussed this in a separate chat over the course of a few days, but I wanted to move the conclusion/solution over here for posterity.

After receiving your files and data (which contained @CPKalb fix for timing issues), I ran against the latest METplus beta and confirmed similar behavior: BRIERCL ranged in values depending on the masking region, and none with a value of 0.25 (as should be the case for a constant climo field and stddev field).

During my testing, I found that by omitting the stddev field but keeping the climo field, the BRIERCL values for all the masking regions became closer to 0.25 (but still not quite there). That was because of how MET evaluates those fields when looking at probabilities. The following is from the MET user's guide:

> When “climo_mean” is set to a probability field with a range of [0, 1] and “climo_stdev” is unset, the MET tools use the “climo_mean” probability values directly to compute Brier Skill Score (BSS).

Before that, when both the climo and stddev fields were provided, MET was using the following logic:

> When “climo_mean” and “climo_stdev” are both set to non-probability fields, the MET tools use the mean, standard deviation, and observation event threshold to derive a normal approximation of the climatological probabilities.

Because the normal approximation didn't match the single value field at all, those BRIERCL values were much further away.

But as mentioned before, even with this fix, the BRIERCL was never quite at 0.25.

The solution came from @JohnHalleyGotway, who saw that the thresholds being used were interfering with how to properly bin the climatology data. In the configuration file, the `cat_thresh` was set to ==0.1, creating 10 bins with the start point for each bin at the 0.1 increment. Because the 0.5 bin starts at exactly 0.50000000, having a constant climo field of 0.5's caused a problem: sometimes, a gridpoint would be interpreted as 0.5000000001, and would fall to the right of the bin edge. And other times, a gridpoint would be interpreted as 0.4999999999, and fall to the left of the bin edge. So to correct this, all that's needed is a change of where the bin edge is. Instead of `cat_thresh = ==0.1`, use a `cat_thresh= ==0.2` and you should see BRIERCL's of 0.25 for all masking regions.

This is all related to how MET calculates probability scores. Instead of looking at the value itself and determining the Brier score (as we would do with a pencil and paper approach), MET creates an nx2 contingency table and determines where each gridpoint falls, then calculates the appropriate scores. So it's very important to see how thresholds are set up and what the data will do that resides on an edge of those thresholds/bins.
